### ISSUE TITLE
Type declaration `parent` is erroneously detected as `self`

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#1681]
```php
<?php

class A
{
    public function returnSelf() : self
    {
        return $this;
    }
}

class B extends A
{
    public function returnSelf() : parent
    {
        return parent::returnSelf();
    }
}
```

https://psalm.dev/r/1a2dc56f88 :

```
Psalm output (using commit 9fb9679):

ERROR: MethodSignatureMismatch - 13:5 - Method B::returnSelf with return type 'B' is different to return type 'A' of inherited method A::returnSelf

INFO: LessSpecificReturnStatement - 15:16 - The type 'A' is more general than the declared return type 'B' for B::returnSelf

INFO: MoreSpecificReturnType - 13:36 - The declared return type 'B' for B::returnSelf is more specific than the inferred return type 'A'
```
